File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    
    const genreMap = new Map();
    
    for(let i=0; i<genres.length; i++){
        genreMap.set(genres[i], {total: 0, songs: []});
    }  
    
    for(let i=0; i<genres.length; i++){
        const g = genreMap.get(genres[i]);
        if(!g){
            genreMap.set(genres[i], {total: plays[i], songs: [{id: i, play: plays[i]}]});
        }
        
        else{
            g.total += plays[i];
            g.songs.push({id: i, play: plays[i]});
        }
    }
    
    const genreArr = [...genreMap].sort((a,b)=>b[1].total - a[1].total)
    
    genreArr.forEach((genre, idx)=>{
        genre[1].songs.sort((a,b)=> b.play - a.play)
        
        genre[1].songs.forEach((s,idx)=>{
            if(idx<2){
                answer.push(s.id);
            }
        })

    })
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    
    genre_kinds = defaultdict(int)
    albums = defaultdict(list)
    
    for i, g in enumerate(genres):
        genre_kinds[g] += plays[i]
        albums[g].append((i, plays[i]))
    
    genre_kinds = sorted(genre_kinds.keys(), key= lambda x: genre_kinds[x], reverse = True)
    
    for genre in genre_kinds:
        album = sorted(albums[genre], key= lambda x: (-x[1], x[0]))
        
        for idx, a in enumerate(album):
            i, album = a
            if idx > 1:
                break;
            answer.append(i)
    
    
        
    return answer